pallene/parser.lua

```diff
@@ -1055,16 +1086,43 @@ function Parser:unexpected_token_error(non_terminal)
 end
 
 function Parser:wrong_token_error(expected_name, open_tok)
-    local loc   = self.next.loc
+
+    local next_tok = self.next
+    local is_stolen_delimiter = false
+
+    if open_tok then
+        for i = 1, #self.mismatched_indentation, 2 do
+            local susp_open  = self.mismatched_indentation[i]
+            local susp_close = self.mismatched_indentation[i+1]
+            if expected_name == susp_close.name and susp_open.loc.pos > open_tok.loc.pos then
+                open_tok = susp_open
+                next_tok = susp_close
+                is_stolen_delimiter = true
+                break
+            end
+        end
+    end
+
+    local loc = next_tok.loc
     local what  = self:describe_token_name(expected_name)
-    local where = self:describe_token(self.next)
+    local where = self:describe_token(next_tok)
+
     if not open_tok or loc.line == open_tok.loc.line then
         self:syntax_error(loc, "expected %s before %s", what, where)
     else
         local owhat = self:describe_token_name(open_tok.name)
-        self:syntax_error(loc, "expected %s before %s, to close the %s at line %d",
-            what, where, owhat, open_tok.loc.line)
+        local oline = open_tok.loc.line
+        if is_stolen_delimiter then
+            self:syntax_error(loc,
+                "expected %s to close %s at line %d, before this less indented %s",
```

Comment:
Is it easy to tell what "**this** less indented token" refers to when reading the error?
Ideally, we'd be highlighting the exact places on the terminal, but Pallene does not need that atm so I guess this the best we can do.